src/main/java/io/vertx/core/impl/VertxImpl.java

```diff
@@ -891,10 +893,11 @@ public void run() {
 
     public void handle(Void v) {
```

Comment:
I suppose this is going to happen in the event loop, but `cancel` can happen from any thread.
If that's true I can see this case to happen:
- thread 1 schedule a periodic task on the event loop 
- the event loop is ready to fire it the first time and `disposed::get` return `false`: GC pause the event loop before calling `handler::handle(timerID)`
- thread N try cancel the timer and win the race cancelling it
- the event loop is resumed (damn GC!) and it's calling `handler::handle` on the already closed/racing closing task

If this scenario is fine to happen, I don't need to propose an alternative solution to this
ie using an hand-rolled spin lock-like shared int that takes care of tracking the handling state (handling -> idle) in a way that a `tryCancel` can decide to spin await (or give up, returning false) and `compareAndSet(idle, disposed)`.
And any update to the concurrent hash map could/should depend by the ability of correctly setting a new state

Reply:
it can happen from any thread

Reply:
cancelling a timer can fail (it returns a boolean), what is not fine is to have the task executed and have cancel return true

Reply:
I think the disposed.get() is OK because the timer can fire several times and cancel will always return true.

the bug was with the other path with the CAS

Reply:
For me this scenario is fine. 
I do not expect being able to cancel a timer that has fired and has started running.

For me when `disposed::get` has been called, I consider that the time to run the timer has come and that we have commited to starting it -or- that it has already started.